### DISCUSSION TITLE
anchor positon configuration

### DISCUSSION THREAD

**Discussion by dongxueyuu** [bitcraze/discussions#741] [Q&A]
hello everyone!
In the tutorials of getting started with the LPS,when I add new anchor and enter the position of anchors in the client and click the button "write to anchors", nothing happens no matter how many times I retry or restart,the position won't be green or red,they are grey.What's the reason of that?
thank you.

**Comment by knmcguire** [comment]
Hi! We need some more information!

1- Which OS are you using, which version of python and the version of the cfclient?
2- Have you updated the anchors as indicated at the start of the [tutorial?](https://www.bitcraze.io/documentation/tutorials/getting-started-with-loco-positioning-system/) 
3- When you press the button, do you see anything strange in the terminal of where you started the cfclient?
4- When you press the button, do you see anything in [the console tab](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/#tabs-and-toolboxes)

**Comment by dongxueyuu** [comment]
Hi, thanks for your answer!
 
1.I use the Bitcraze-VM installed according to the tutorial.The version of cfclient is &nbsp;2022.3rc2 and the Python is 3.8.10.
 
2.Yes,I have set up the loco positioning system exactly as described on the tutorial.
 
3.Nothing strange in the terminal when I press the “write to anchors”,but there are two lines warning after the crazyflie is connected. I don’t know what it means.



4.Nothing appears in the console tab when I press the botton.

I attached the picture below.


 







米頔
***@***.***



&nbsp;




------------------&nbsp;Original&nbsp;------------------
From:                                                                                                                        "bitcraze/discussions"                                                                                    ***@***.***&gt;;
Date:&nbsp;Tue, May 23, 2023 05:30 PM
***@***.***&gt;;
***@***.******@***.***&gt;;
Subject:&nbsp;Re: [bitcraze/discussions] anchor positon configuration (Discussion #741)





 
Hi! We need some more information!
 
1- Which OS are you using, which version of python and the version of the cfclient?
 2- Have you updated the anchors as indicated at the start of the tutorial?
 3- When you press the button, do you see anything strange in the terminal of where you started the cfclient?
 4- When you press the button, do you see anything in the console tab

**Reply by knmcguire** [reply]
For some reason, the pictures are not visible. I believe that you have answered this github discussion through email. Could you add the pictures again but directly on the github website?

**Reply by dongxueyuu** [reply]
Two lines of warning are like this:
![微信图片_20230523193347](https://github.com/bitcraze/discussions/assets/134368713/0569da0b-9e1d-440a-a289-8a442c21bae7)

And the cfclient interface is :
<img width="959" alt="屏幕截图 2023-05-23 195031" src="https://github.com/bitcraze/discussions/assets/134368713/d840391d-1218-4082-9790-53fd6c70c29b">

**Reply by knmcguire** [reply]
I see anchor 1 2 4 and 7 green in the cfclient, so then the crazyflie is able to receive them. And it seems that the crazyflie is able to estimate a position as well.

![image](https://github.com/bitcraze/discussions/assets/11061260/488e72d6-743c-4313-a9b9-8b5717d58c98)
The dialog window where you press write to anchors will not give any 'feedback' that it worked I believe (bad gui design...) but from the crazyflie status it seemed that it worked?

**Reply by dongxueyuu** [reply]
Hello,I have tried to verify the anchor positions according to the tutotial,to see if it's like you say that even there is no feedback but it works.But it doesn't work well in fact,the positions are not accurate.
Besides,I found that when the crazyfile and anchors are connected,the LED on the Loco dack are not on.I don’t know if that’s the reason.If this is possible,could you tell that why the LED on the deck is not on and how to deal it?
Thank you very much.
![微信图片_20230525203856](https://github.com/bitcraze/discussions/assets/134368713/326a56f8-bfd4-46e5-b514-648f7a9fdd9d)

**Reply by krichardsson** [reply]
Hi!
It is strange that the LEDs are not blinking, it sounds like your deck might be broken. Could you please reboot your Crazyflie, connect with the python client and copy/paste the console log (from the console tab in the client) here.

**Reply by dongxueyuu** [reply]
SYS: Build 0:0c52256c6f30 (2022.03) CLEAN
sYs:I am 0x2030364330565009002E002E and i have 1024KB of flash!
CFGBLK: v1, verification 
DECK INFO: CONFIG DECK FORCE=1cMeshDeck found
DECK INFO: compile-time forced driver 1cMeshDeck added
DECK CORE: 3 deck(s) found
DECK CORE: Calling INIT on driver bcDWM1000 for deck 0
DECK CORE: Calling INIT on driver bcFlow2 for deck 1
ZR2: Z-down sensor
PMW: Motion chip id: 0x49:0xB6
DECK CORE: Calling INIT on driver 1cMeshDeck for deck 2
MESHDECK: Mesh deck init
IMU: BMI088: Using 12C interface.
IMU: BMI088 Gyro connection [OKI
IMU: BMI088 Accel connection [OK
IMU: BMP388 12C connection 
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
sys: About to run tests in system.c.
SYS: NRF51 version: NA
EEPROM: 12C connection .
STORAGE: storage check [OK1.
IMU: BMI088 gyro self-test 
DECK CORE: Deck 0 test [OK1.
DECK CORE: Deck 1 est 
DECK CORE: Deck 2 test [OK1.
sys: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 14280 bytes
STAB: Ready to fly.
ESTKALMAN: State out of bounds, resetting
DWM: Automatic mode: detected TWR
<img width="553" alt="屏幕截图 2023-05-26 175010" src="https://github.com/bitcraze/discussions/assets/134368713/5f1a33c5-a83b-4a13-847e-7f77be13ba2a">

**Reply by krichardsson** [reply]
Thanks for the log!

I can see that you also have a "mesh deck" (I don't  know what it is, but sounds cool :-) ) and flow deck. Would it be possible to disconnect all decks except the Loco deck and see if it makes a difference? It would be nice to be able to rule out any possible interference between the decks.

**Reply by dongxueyuu** [reply]
Thanks for your reply very much!
I have re-flashed the crazeflie with an official release and the LEDs on deck are blinking now.I have also disconnected all decks except Loco deck,but there is no diffierence.I still can’t write position to anchors ☹.
![微信图片_20230530212153](https://github.com/bitcraze/discussions/assets/134368713/40e8a25d-38a6-45bc-98cc-18dc7ef46354)

**Reply by krichardsson** [reply]
Good news!
I think this is an indication of that the deck actually works and that it might be a configuration problem.

I will outline how the process of anchor position works to make it easier to discuss possible issues. The anchor positions are stored in the anchors and transmitted together with timing information to the Crazyflie in the UWB packes when the system is running. This means that the Crazyflie will not know the position of an anchor until it has received a packet from it. When setting the anchor position, the Crazyflie sends the new position in an anchor in an UWB packet. When the packet is received in the anchor, it will store the new position and also use it in packets that are transmitted. When a packet with the new position is received in the Crazyflie, the Crazyflie will know about the new position and the process is complete. The client is of course involved as well and the full round trip for setting anchor positions looks like this:
Client ->  Crazyflie -> Anchor -> Crazyflie -> Client

I can see in the picture that anchors 0, 3, 5 and 6 are red which means that they are not currently received.
I can also see in the picture that you have configured positions for anchors 1, 2, 4 and 7. If you can get these positions also after restarting the Crazyflie it means that they have been written to the anchors successfully.
 
It is possible that the UWB communication between anchors and Crazyflie is not working properly. The default transmission power of the UWB radios is fairly low and if the distance is a bit too long many packets are lost. You can test if this is the problem by moving the Crazyflie closer to one of the red anchors and see if it becomes green.
This can also explain why there is a problem writing new positions to the anchors, the packets might get lost when sent from the Crazyflie.

To increase the UWB transmission power in the Crazyflie you must rebuild the firmware with a compile flag. In the firmware repository [enter the configuration tool](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/kbuild/), go to the "Expansion deck configuration" menu and tick "Full TX power". Rebuild and flash the firmware to the Crazyflie.

To increase the UWB transmission power of the anchors, use the [low level configuration](https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/development/anchor-low-level-config/). Set the power to 9 and forceTxPower, this is done by pressing keys `pf9`
